File: tasksync-app/lib/features/auth/domain/value_objects/password.dart
import 'package:app/core/value_object/value_object.dart';

class Password extends ValueObject<String> {
  @override
  final String value;

  const Password(this.value);

  factory Password.empty() => const Password('');

  static final _passwordRegex = RegExp(
    r'^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*()_+\-=\[\]{}|;:",./<>?]).{6,}$',
  );

  @override
  String? get errorMessage {
    if (value.isEmpty) {
      return 'Password is required';
    }
    if (value.length < 6) {
      return 'Password must be at least 6 characters';
    }
    if (!_passwordRegex.hasMatch(value)) {
      return 'Must include uppercase, lowercase, number & special char';
    }
    return null;
  }
}

File: tasksync-api/internal/core/objects/password.go
package objects

import (
	"errors"
	"strings"
)

var (
	ErrPasswordRequired  = errors.New("password is required")
	ErrPasswordMinLength = errors.New("password must be at least 6 characters")
	ErrPasswordMaxLength = errors.New("password must be at most 255 characters")
	ErrPasswordInvalid   = errors.New("password must contain at least one uppercase letter, one lowercase letter, one number, and one special character")
)

type Password struct {
	value string
}

func NewPassword(value string) (Password, error) {
	value = strings.TrimSpace(value)

	if value == "" {
		return Password{}, ErrPasswordRequired
	}

	if len(value) < 6 {
		return Password{}, ErrPasswordMinLength
	}

	if len(value) > 255 {
		return Password{}, ErrPasswordMaxLength
	}

	return Password{value: value}, nil
}

func (p Password) Value() string {
	return p.value
}
